doc/developer/design/20230216_role_based_access_control.md

```diff
@@ -0,0 +1,357 @@
+
+Here is a summary of all the privileges, attributes, and ownership needed to perform certain actions:
+
+| Operation                            | Privilege, Attribute, and OwnerShip                                         |
+|--------------------------------------|-----------------------------------------------------------------------------|
+| `ALTER` (NOT role)                   | Ownership of the object, `SCHEMA(C)`                                        |
+| `ALTER ROLE`                         | `CREATEROLE` (NOTE: Only `SUPERUSER` can change the `SUPERUSER` attribute). |
+| `COPY TO`                            | `CLUSTER(U)`, `OBJECT(r)`                                                   |
+| `COPY FROM`                          | `OBJECT(a)`                                                                 |
+| `CREATE CLUSTER`                     | `CREATECLUSTER`                                                             |
+| `CREATE CLUSTER REPLICA`             | `CLUSTER(C)`, `CREATECLUSTER`                                               |
+| `CREATE {SECRET, TABLE, TYPE, VIEW}` | `SCHEMA(C)`                                                                 |
+| `CREATE CONNECTION`                  | `SCHEMA(C)`, sometimes `SECRET(U)`                                          |
+| `CREATE DATABASE`                    | `CREATEDATABASE`                                                            |
+| `CREATE INDEX`                       | `SCHEMA(C)`, `CLUSTER(C)`                                                   |
+| `CREATE MATERIALIZED VIEW`           | `CREATEPERSIST`, `SCHEMA(C)`, `CLUSTER(C)`                                  |
+| `CREATE SOURCE`                      | `CREATEPERSIST`, `SCHEMA(C)`, `CLUSTER(C)` sometimes `CONNECTION(U)`        |
+| `CREATE TABLE`                       | `CREATEPERSIST`, `SCHEMA(C)`                                                |
+| `CREATE {ROLE, USER}`                | `CREATEROLE` (NOTE: only `SUPERUSER`s can create other `SUPERUSER`s).       |
+| `CREATE SCHEMA`                      | `DATABASE(C)`                                                               |
+| `DELETE`                             | `OBJECT(d)` usually `CLUSTER(U)`, `OBJECT(r)`                               |
+| `DROP` (NOT ROLE)                    | Ownership of the object                                                     |
+| `DROP ROLE`                          | `CREATEROLE` (NOTE: only `SUPERUSER`s can drop other `SUPERUSER`s).         |
+| `EXPLAIN`                            | usually `OBJECT(r)`                                                         |
+| `INSERT INTO ... VALUES`             | `OBJECT(a)`                                                                 |
+| `INSERT INTO ... SELECT`             | `CLUSTER(U)`,`OBJECT(a)` usually `OBJECT(r)`                                |
+| `{SELECT, SHOW, SUBSCRIBE}`          | `CLUSTER(U)`, usually `OBJECT(r)`                                           |
```

Comment:
As Cara pointed out, this could prevent `SHOW` and other introspection queries which might not be desired. Do we want to special case those somehow? This one is tricky to get right.

Reply:
I have one idea that has been shot down before, but might warrant reconsideration in this context. All `SHOW` commands are run in the `mz_introspection` cluster, and all roles are automatically given `CLUSTER(U)` for `mz_introspection` and `OBJECT(r)` for builtin objects.

Reply:
It always seemed a little strange to make users hop clusters to run introspection queries, so that option is worth considering for ux alone. 

Flagging that there are some `mz_introspection` use cases we missed in the initial spec, which I'm adding there too:
- Support uses `mz_introspection` clusters to debug client environments 
- They need to be able to to SELECT across all mz_* tables and execute EXPLAIN and DESCRIBE, ideally across clusters
- Current permissions on this cluster are hardcoded in environmentd?

Reply:
What about select queries on objects that don't inherently exist on a cluster?
Would you need to be in a cluster that you have `CLUSTER(U)` privileges for that cluster when you run the `SELECT`?

Aka we're saying that anyone who has access to Materialize should have usage access to at least one cluster.

Reply:
> Would you need to be in a cluster that you have CLUSTER(U) privileges for that cluster when you run the SELECT?

For this design yes. The rationale is that your select will use some of the compute resources in the current cluster, and therefore you will need usage privileges on that cluster.